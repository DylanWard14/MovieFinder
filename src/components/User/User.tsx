import React, { Fragment } from 'react';
import styled from 'styled-components/macro';

import { UserAuthState } from '../../store/User-auth/types';
import { UserDetailsState } from '../../store/User-details/types';
import UserSkeleton from '../../skeletons/UserSkeleton/UserSkeleton';
import Button from '../Button/Button';

const Avatar = styled.img`
    width: 2rem;
    height: 2rem;
    border: 2px solid transparent;
    border-radius: 50%;
    margin-left: 0.5rem;
    vertical-align: middle;
`;

const Username = styled.p`
    display: inline-block;
    margin: 0;
`;

export interface UserProps {
    userAuth: UserAuthState;
    userDetails: UserDetailsState;
    handleLogin: () => void;
}

const User: React.FunctionComponent<UserProps> = ({ userAuth, userDetails, handleLogin }) => {
    if (userAuth.loading || userDetails.loading || (userAuth.data?.authorized && !userDetails.data)) {
        return <UserSkeleton />;
    }
    if (!userAuth.data?.authorized || !userDetails.data) {
        return <Button onClick={handleLogin}>Login</Button>;
    }
    return (
        <Fragment>
            <Username>{userDetails.data.username}</Username>
            <Avatar src={`https://www.gravatar.com/avatar/${userDetails.data.avatar.gravatar.hash}`} />
        </Fragment>
    );
};

export default User;
