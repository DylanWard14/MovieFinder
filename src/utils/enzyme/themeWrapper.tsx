import React from 'react';
import { render, shallow, mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';

import myTheme from '../../theme';

export const renderWrapper = (component: React.ReactNode) => {
    return render(<ThemeProvider theme={myTheme}>{component}</ThemeProvider>);
};

export const shallowWrapper = (component: React.ReactNode) => {
    return shallow(<ThemeProvider theme={myTheme}>{component}</ThemeProvider>);
};

export const mountWrapper = (component: React.ReactNode) => {
    return mount(<ThemeProvider theme={myTheme}>{component}</ThemeProvider>);
};
