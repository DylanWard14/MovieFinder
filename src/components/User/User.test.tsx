import React from 'react';
import { renderWrapper, mountWrapper } from '../../utils/enzyme/themeWrapper';

import User from './User';

describe('User', () => {
    const handleLogin = jest.fn();
    const setup = (propsOverrides = {}) => {
        const props = {
            userAuth: {
                loading: false,
                data: {
                    authorized: true,
                },
            },
            userDetails: {
                loading: false,
                data: {
                    avatar: {
                        gravatar: {
                            hash: 'abc123',
                        },
                    },
                    id: 123,
                    iso_639_1: '123',
                    iso_3166_1: '123',
                    name: 'Dylan Ward',
                    include_adult: false,
                    username: 'Dward',
                },
            },
            handleLogin: handleLogin,
            ...propsOverrides,
        };

        return {
            rendered: renderWrapper(<User {...props} />),
            mounted: mountWrapper(<User {...props} />),
        };
    };
    describe('snapshot', () => {
        it('renders correctly', () => {
            const { rendered } = setup({});

            expect(rendered).toMatchSnapshot();
        });
    });

    describe('mounted', () => {
        it('renders loading state when userAuth is loading', () => {
            const { mounted } = setup({ userAuth: { loading: true } });

            expect(mounted.find('UserSkeleton').exists()).toBeTruthy();
        });

        it('renders loading state when userDetails is loading', () => {
            const { mounted } = setup({ userDetails: { loading: true } });

            expect(mounted.find('UserSkeleton').exists()).toBeTruthy();
        });

        it('renders loading state when user is authorized but no userDetails data is loading', () => {
            const { mounted } = setup({ userAuth: { data: { authorized: true } }, userDetails: { data: undefined } });

            expect(mounted.find('UserSkeleton').exists()).toBeTruthy();
        });

        it('does not render loading state when the is userAuth and userDetails data', () => {
            const { mounted } = setup();

            expect(mounted.find('UserSkeleton').exists()).toBeFalsy();
        });

        it('renders the Username and avatar correctly', () => {
            const { mounted } = setup();

            expect(mounted.find('p').text()).toBe('Dward');
            expect(mounted.find('img').prop('src')).toBe('https://www.gravatar.com/avatar/abc123');
        });

        it('renders the login button when the user is not authorized and there is no userDetails data', () => {
            const { mounted } = setup({ userAuth: { data: { authorized: false } }, userDetails: { data: undefined } });

            expect(mounted.find('button').exists()).toBeTruthy();
        });

        it('handleLogin is called when button is clicked', () => {
            const { mounted } = setup({ userAuth: { data: { authorized: false } }, userDetails: { data: undefined } });

            expect(handleLogin).toHaveBeenCalledTimes(0);
            mounted.find('button').simulate('click');
            expect(handleLogin).toHaveBeenCalled();
        });
    });
});
