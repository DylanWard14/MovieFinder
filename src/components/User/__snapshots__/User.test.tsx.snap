// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User snapshot renders correctly 1`] = `
Array [
  <p
    class="User__Username-sc-1j5wnkp-1 kSkEXl"
  >
    Dward
  </p>,
  <img
    class="User__Avatar-sc-1j5wnkp-0 ipnUnN"
    src="https://www.gravatar.com/avatar/abc123"
  />,
]
`;
